### PR TITLE
[CHORE] Remove CountMode and ResourceRequest from public API

### PR DESCRIPTION
**Pull request by desmondcheongzx**
CountMode is used in [`Expression.count`](https://www.getdaft.io/projects/docs/en/latest/api_docs/doc_gen/expression_methods/daft.Expression.count.html), but it's not user friendly to have users figure out how to use a CountMode object. Instead we get public methods to accept countmode strings instead.

ResourceRequest is used in [DataFrame.with_column](https://www.getdaft.io/projects/docs/en/latest/api_docs/doc_gen/dataframe_methods/daft.DataFrame.with_column.html) and [DataFrame.with_columns](https://www.getdaft.io/projects/docs/en/latest/api_docs/doc_gen/dataframe_methods/daft.DataFrame.with_columns.html) but these are deprecated from public use for the same reasons as CountMode. We thus remove ResourceRequest from the dataframe methods.

### CODE REVIEW

**Review by jaychia — APPROVED** [review]
LGTM, note warnings though.

Do resource requests make it on the public API anywhere now? I think they are primarily used in UDFs, but also believe that the API there is mostly `num_cpus/gpus=...`.

**Comment by desmondcheongzx** [comment]
> Do resource requests make it on the public API anywhere now?
@jaychia they pop up in [DataFrame.with_column](https://www.getdaft.io/projects/docs/en/latest/api_docs/doc_gen/dataframe_methods/daft.DataFrame.with_column.html) and [DataFrame.with_columns](https://www.getdaft.io/projects/docs/en/latest/api_docs/doc_gen/dataframe_methods/daft.DataFrame.with_columns.html)

**Comment by desmondcheongzx** [comment]
As discussed offline, we now favour removing these classes from public API use rather than documenting them.

File: daft/expressions/expressions.py [hunk 1]
```diff
@@ -3010,15 +3011,17 @@ def value_counts(self) -> Expression:
         """
         return Expression._from_pyexpr(native.list_value_counts(self._expr))
 
-    def count(self, mode: CountMode = CountMode.Valid) -> Expression:
+    def count(self, mode: str | CountMode = CountMode.Valid) -> Expression:
```
Comment (jaychia) [inline daft/expressions/expressions.py:3014]: Use `Literal["all", "valid", "null"]` instead to be more specific about what values can be passed in

**Review by jaychia — APPROVED** [review]

File: daft/expressions/expressions.py (hunk 1 above)
Comment (desmondcheongzx) [reply daft/expressions/expressions.py]: Sounds good, done

**Pull request merged**